### PR TITLE
[WIP] Custom Policies Front

### PR DESCRIPTION
**Pull request by didierofrivia**
**What this PR does / why we need it**:

New feature: Custom Policies.

This is the frontend part of the new feature about crafting new policies.

**Which issue(s) this PR fixes** 

**Verification steps** 

**Special notes for your reviewer**:

### CODE REVIEW

File: app/javascript/src/Policies/styles/policies.scss [hunk 1]
```diff
@@ -11,13 +11,16 @@ $subtleColor: #ddd;
 
 .PoliciesWidget {
   position: relative;
-  width: 510px;
-  margin: 0 auto;
   line-height: 1.5;
   padding-bottom: 2rem;
 
+  .PolicyChain, .PolicyRegistryList, .PolicyConfiguration {
+    width: 510px;
```
Comment (thomasmaas) [inline app/javascript/src/Policies/styles/policies.scss]: pxs are forbidden where i come from

Comment (didierofrivia) [reply app/javascript/src/Policies/styles/policies.scss]: I've just moved it from here: https://github.com/3scale/porta/pull/602/files/bb11cd4c3455502468b1011e9c3c4a784123f9f4..331a9c841cc5ed34614ec7cd9c68c2d4f310e148#diff-dc7eadc737110a7b487300a9ec5b4f00L14

But better if you or @damianpm polish the styling afterwards

File: app/controllers/admin/api/registry/policies_controller.rb [hunk 2]
```diff
@@ -53,19 +53,15 @@ def show
...
-    policy_params = params.require(:policy)
-    final_params = policy_params.permit(:name, :version)
-    final_params.merge(schema: policy_params.require(:schema)).permit!
+    PermittedParams::PolicyParams.new(params.require(:policy)).to_params
   end
 
-  def find_policy
-    @policy ||= current_account.policies.find_by_id_or_name_version!(params[:id])
+  def policy
+    @policy ||= current_account.policies.find_by!(id_or_name_version: params[:id])
```
Comment (guicassolato) [inline app/controllers/admin/api/registry/policies_controller.rb]: I don't think this works.

File: app/controllers/provider/admin/registry/policies_controller.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+  layout 'provider'
+
+  def index
+    @policies = Policies::PoliciesListService.call(current_account, builtin: false)
+  end
+
+  def create
+    policy = current_account.create(create_policy_params)
+    @policy = Policies::PoliciesListService.new
```
Comment (guicassolato) [inline app/controllers/provider/admin/registry/policies_controller.rb]: If we're going to use the service, maybe rename the variable, as it does not return really a "policy". WDYT?

File: app/javascript/src/Policies/components/CustomPolicy.jsx.orig [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+      } catch (err) {
+        setState({ valid: false, code })
+      }
+    })
+  }
+
+  const icon = state.valid ? 'check' : 'times'
+  const cls = state.valid ? 'valid' : 'invalid'
+  const hiddenClass = state.valid ? 'is-hidden' : ''
+  const errorClass = state.valid ? '' : 'CustomPolicy-error'
```
Comment (guicassolato) [inline app/javascript/src/Policies/components/CustomPolicy.jsx.orig]: Can we use destructuring assignment to check for `state.valid` only once? Or not a good idea due to browser support? Or not too complicated?

```javascript
const [icon, cls, hiddenClass, errorClass] = state.valid ? ['check', 'valid', 'is-hidden', ''] : ['times', 'invalid', '', 'CustomPolicy-error'];
```

File: app/javascript/src/Policies/components/CustomPolicy.jsx.orig [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+      <div className="CustomPolicy-editor">
+        <Editor className="CustomPolicy-code" code={pol.configuration} onChange={onSchemaEdited(pol)} />
+        <SchemaForm className="CustomPolicy-form" schema={pol.configuration} />
+      </div>
+      <form action={`/p/admin/registry/policies/${pol.name}-${pol.version}`} method="post">
+        <input type="text" name="name" value={pol.name} onChange={handleChange(pol)} />
+        <input type="text" name="version" value={pol.version} onChange={handleChange(pol)} />
+        <textarea name="summary" id="" cols="30" rows="10" value={pol.summary} onChange={handleChange(pol)} />
+        <textarea name="description" id="" cols="30" rows="10" value={pol.description} onChange={handleChange(pol)} />
+<<<<<<< Updated upstream
```
Comment (guicassolato) [inline app/javascript/src/Policies/components/CustomPolicy.jsx.orig]: oops. Attention here.

File: app/javascript/src/Policies/components/CustomPolicy.jsx [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+      setState({ valid: true, code })
+    } catch (err) {
+      setState({ valid: false, code })
+    }
+  }
+
+  const icon = state.valid ? 'check' : 'times'
+  const cls = state.valid ? 'valid' : 'invalid'
+  const hiddenClass = state.valid ? 'is-hidden' : ''
+  const errorClass = state.valid ? '' : 'CustomPolicy-error'
```
Comment (guicassolato) [inline app/javascript/src/Policies/components/CustomPolicy.jsx]: Can we use destructuring assignment to check for `state.valid` only once? Or not a good idea due to browser support? Or too complicated?

```javascript
const [icon, cls, hiddenClass, errorClass] = state.valid ? ['check', 'valid', 'is-hidden', ''] : ['times', 'invalid', '', 'CustomPolicy-error'];
```

File: test/unit/permitted_params/policy_params_test.rb [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    assert_equal({name: 'hello', version: '1.0', schema: 'schema'}.stringify_keys, params.to_params)
+  end
+
+  test 'schema is ommitted' do
+    params = PermittedParams::PolicyParams.new(name: 'hello', version: '1.0')
+    assert_equal({name: 'hello', version: '1.0', schema: nil}.stringify_keys, params.to_params)
+  end
+
+  test 'schema is a Hash' do
+    params = PermittedParams::PolicyParams.new(schema: {foo: :bar})
```
Comment (guicassolato) [inline test/unit/permitted_params/policy_params_test.rb]: If schema is a stringified json object, will it be accessible as a hash?

File: app/controllers/provider/admin/registry/policies_controller.rb (hunk 3 above)
Comment (didierofrivia) [reply app/controllers/provider/admin/registry/policies_controller.rb]: you are right, we were trying to mimic APIcast behaviour... but probably doesn't make sense here

File: app/javascript/packs/custom_policy.js [hunk 8]
```diff
@@ -0,0 +1,12 @@
+import {render} from 'react-dom'
+import React from 'react'
+import {CustomPolicy} from 'Policies/components/CustomPolicy'
+import {safeFromJson, parsePolicies} from 'Policies/util'
+
+document.addEventListener('DOMContentLoaded', () => {
+  const container = document.getElementById('custom-policy-container')
+  const jsonPolicy = container.dataset.policy
+  // TODO: Notify the user something has failed. Error messages.
+  const props = (jsonPolicy) ? {policy: parsePolicies(safeFromJson(jsonPolicy, () => []))[0]} : {}
```
Comment (didierofrivia) [inline app/javascript/packs/custom_policy.js]: second iteration we need to normalize the way we get policies from different sources and how we store them.

File: app/javascript/src/Policies/components/CustomPolicy.jsx [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+import SchemaForm from 'react-jsonschema-form'
+import { fromJson, toJson } from 'Policies/util'
+import type { RegistryPolicy } from 'Policies/types/Policies'
+import type {InputEvent} from 'Policies/types'
+import 'codemirror/mode/javascript/javascript'
+import 'Policies/styles/policies.scss'
+
+type OnChange = (InputEvent) => void
+
+function CSRFToken ({win = window}: {win?: any}): React.Node {
```
Comment (didierofrivia) [inline app/javascript/src/Policies/components/CustomPolicy.jsx]: There's no flow type for Window.

**Comment by didierofrivia** [comment]
Closed in favour of https://github.com/3scale/porta/pull/687

**Pull request closed without merging**